enterprise/server/remote_execution/filecache/filecache.go

```diff
@@ -229,19 +248,20 @@ func (c *fileCache) FastLinkFile(node *repb.FileNode, outputPath string) (hit bo
 	return true
 }
 
-func (c *fileCache) AddFile(node *repb.FileNode, existingFilePath string) error {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-
+func (c *fileCache) addFileToGroup(groupID string, node *repb.FileNode, existingFilePath string) error {
 	// Remove any existing entry. We can't update in-place because if we
 	// overwrite an existing link with different contents, all pointers
 	// to the old link would suddenly change to point to the new content,
 	// which is not good.
-	k := key(node)
+	k := groupSpecificKey(groupID, node)
 	c.l.Remove(k)
 
-	fp := c.filecachePath(node)
+	fp := c.filecachePath(k)
+	if err := disk.EnsureDirectoryExists(filepath.Dir(fp)); err != nil {
```

Comment:
Any worry in adding an additional `stat` call for each fast link for actions with tens of thousands of input files instead of doing it once up front?

Reply:
I did not notice an impact when testing this locally, so decided to leave this as-is for simplicity.